letsencrypt/client/auth_handler.py

```diff
@@ -259,156 +259,99 @@ def _challenge_factory(self, domain, path):
         :param list path: List of indices from `challenges`.
 
         :returns: dv_chall, list of
-            :class:`letsencrypt.client.challenge_util.IndexedChall`
+            :class:`letsencrypt.client.achallenges.Indexed`
             client_chall, list of
-            :class:`letsencrypt.client.challenge_util.IndexedChall`
+            :class:`letsencrypt.client.achallenges.Indexed`
         :rtype: tuple
 
         :raises errors.LetsEncryptClientError: If Challenge type is not
             recognized
 
         """
-        challenges = self.msgs[domain].challenges
-
         dv_chall = []
         client_chall = []
 
         for index in path:
-            chall = challenges[index]
-
-            # Authenticator Challenges
-            if chall["type"] in constants.DV_CHALLENGES:
-                dv_chall.append(challenge_util.IndexedChall(
-                    self._construct_dv_chall(chall, domain), index))
-
-            # Client Challenges
-            elif chall["type"] in constants.CLIENT_CHALLENGES:
-                client_chall.append(challenge_util.IndexedChall(
-                    self._construct_client_chall(chall, domain), index))
+            chall = self.msgs[domain].challenges[index]
+
+            if isinstance(chall, challenges.DVSNI):
+                logging.info("DVSNI challenge for %s.", domain)
+                achall = achallenges.DVSNI(
+                    chall=chall, domain=domain, key=self.authkey[domain])
+            elif isinstance(chall, challenges.SimpleHTTPS):
+                logging.info("SimpleHTTPS challenge for %s.", domain)
+                achall = achallenges.SimpleHTTPS(
+                    chall=chall, domain=domain, key=self.authkey[domain])
+            elif isinstance(chall, challenges.DNS):
+                logging.info("DNS challenge for %s.", domain)
```

Comment:
Minor: identation before DNS challenge were intentional. I believe it breaks up the monotony and makes it easier to read the output.
Performing the following challenges...
    DVSNI
    RecoveryToken
...